pkgs/jnigen/java/src/main/java/com/github/dart_lang/jnigen/apisummarizer/disasm/AsmMethodVisitor.java

```diff
@@ -27,27 +28,66 @@ public void visitParameter(String name, int access) {
   }
 
   @Override
-  public void addAnnotation(JavaAnnotation annotation) {
+  public AnnotationVisitor visitAnnotation(String descriptor, boolean visible) {
+    var annotation = TypeUtils.annotationFromDescriptor(descriptor);
     method.annotations.add(annotation);
-  }
-
-  @Override
-  public AnnotationVisitor visitAnnotationDefault(String descriptor, boolean visible) {
-    return AsmAnnotatedElementVisitor.super.visitAnnotationDefault(descriptor, visible);
+    return new AsmAnnotationVisitor(annotation);
   }
 
   @Override
   public AnnotationVisitor visitTypeAnnotation(
       int typeRef, TypePath typePath, String descriptor, boolean visible) {
-    // TODO(#23): Collect annotation on type parameter
+    var typeReference = new TypeReference(typeRef);
+    var annotation = TypeUtils.annotationFromDescriptor(descriptor);
+    TypeUsage.ReferredType startingType = null;
+    switch (typeReference.getSort()) {
+      case TypeReference.METHOD_TYPE_PARAMETER:
+        {
+          var index = typeReference.getTypeParameterIndex();
+          method.typeParams.get(index).annotations.add(annotation);
+          break;
+        }
+      case TypeReference.METHOD_TYPE_PARAMETER_BOUND:
+        {
+          var typeParamIndex = typeReference.getTypeParameterIndex();
+          var index = typeReference.getTypeParameterBoundIndex();
+          startingType = method.typeParams.get(typeParamIndex).bounds.get(index).type;
+          break;
+        }
+      case TypeReference.METHOD_RETURN:
+        {
+          startingType = method.returnType.type;
+          break;
+        }
+      case TypeReference.METHOD_RECEIVER:
+        {
+          // Not sure what this is.
```

Comment:
Assuming they're using the same jargon as Dart and other languages, this would be the class that owns the method.

Reply:
Yeah. I just wasn't sure how you can annotate the receiver at this point! Like if you have void foo() in some class Foo, where do you put the annotation for it to be associated with class Foo but only for this one method. Maybe Java has something but I couldn't find anything from the docs.